### PR TITLE
Support Xctest for Hydra Lab

### PR DESCRIPTION
**Pull request by lifesaver0129**
Provide XCTest for Hydra Lab.

image

image

image

image

### CODE REVIEW

File: agent/src/main/java/com/microsoft/hydralab/agent/runner/xctest/XctestRunner.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+            commFormat += " -destination \"id=%s\" -resultBundlePath %s";
+            String command = String.format(commFormat, deviceId, resultPath);
+
+            String result = ShellUtils.execLocalCommandWithResult(command, logger);
+            return Const.TaskResult.SUCCESS;
+        } catch (Exception e) {
+            if (logger != null) {
+                logger.error(e.getMessage(), e);
+            }
+            return e.getMessage();
```
Comment (hydraxman) [inline agent/src/main/java/com/microsoft/hydralab/agent/runner/xctest/XctestRunner.java:92]: Throw this

File: agent/src/main/java/com/microsoft/hydralab/agent/runner/xctest/XctestRunner.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Map;
+
+public class XctestRunner extends TestRunner {
```
Comment (hydraxman) [inline agent/src/main/java/com/microsoft/hydralab/agent/runner/xctest/XctestRunner.java:24]: Should this be `XctestRunner` or `XCTestRunner` according to https://developer.apple.com/documentation/xctest

File: agent/src/main/java/com/microsoft/hydralab/agent/runner/xctest/XctestRunner.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    private String runXctest(DeviceInfo deviceInfo, String scope, String suiteName,
+                             String testPkgName, String testRunnerName, Logger logger,
+                             IShellOutputReceiver receiver, TestRun testRun,
+                             int testTimeOutSec, Map<String, String> instrumentationArgs) {
+        if (deviceInfo == null) {
+            throw new RuntimeException("No such device: " + deviceInfo);
+        }
+        if (testTimeOutSec <= 0) {
+            // the test should not last longer than
+            testTimeOutSec = 45 * 60;
```
Comment (hydraxman) [inline agent/src/main/java/com/microsoft/hydralab/agent/runner/xctest/XctestRunner.java:66]: Define this in Const.

File: agent/src/main/java/com/microsoft/hydralab/agent/runner/xctest/XctestRunner.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        }
+        StringBuilder argString = new StringBuilder();
+        if (instrumentationArgs != null && !instrumentationArgs.isEmpty()) {
+            instrumentationArgs.forEach((k, v) -> argString.append(" ").append(k).append(" ").append(v));
+        }
+        String commFormat;
+        if (StringUtils.isBlank(argString.toString())) {
+            commFormat = "xcodebuild test-without-building";
+        } else {
+            commFormat = "xcodebuild test-without-building" + argString;
```
Comment (hydraxman) [inline agent/src/main/java/com/microsoft/hydralab/agent/runner/xctest/XctestRunner.java:76]: What's the dependency of this command? Any environmental requirement changes please clarify it in the agent set up guide: https://github.com/microsoft/HydraLab/wiki/Deploy-a-test-agent-service#ios-mac-agent

File: agent/src/main/java/com/microsoft/hydralab/agent/runner/xctest/XctestRunner.java (hunk 1 above)
Comment (lifesaver0129) [reply agent/src/main/java/com/microsoft/hydralab/agent/runner/xctest/XctestRunner.java:86]: Removed

File: agent/src/main/java/com/microsoft/hydralab/agent/runner/xctest/XctestRunner.java (hunk 2 above)
Comment (lifesaver0129) [reply agent/src/main/java/com/microsoft/hydralab/agent/runner/xctest/XctestRunner.java:23]: Updated

File: common/src/main/java/com/microsoft/hydralab/common/util/Const.java [hunk 5]
```diff
@@ -73,6 +73,9 @@ interface SmartTestConfig {
         String REQUIRE_FILE_NAME = "requirements.txt";
         String STRING_FOLDER_NAME = "SmartTestString";
         String STRING_FILE_NAMES = "strings,username,password";
+        String XCTEST_ZIP_FILE_NAME = "Xctest.zip";
+        String XCTEST_ZIP_FOLDER_NAME = "Xctest";
+        String XCTEST_RESULT_FILE_NAME = "XctestResult";
```
Comment (zhou9584) [inline common/src/main/java/com/microsoft/hydralab/common/util/Const.java:78]: SmartTestConfig interface is used for smarttest. Maybe you can add an interface?

Comment (lifesaver0129) [reply common/src/main/java/com/microsoft/hydralab/common/util/Const.java]: Resolved

File: common/src/main/java/com/microsoft/hydralab/common/util/PkgUtil.java [hunk 6]
```diff
@@ -194,6 +228,87 @@
...
+            //define object
+            ZipEntry entry = null;
+            String entryName = null;
+            String[] names = null;
+            int length;
+            //loop reading entries
+            while (zipEnum.hasMoreElements()) {
+                //get the current entry
+                entry = zipEnum.nextElement();
+                entryName = new String(entry.getName());
```
[the remark this answers is not in the document]
Comment (lifesaver0129) [reply common/src/main/java/com/microsoft/hydralab/common/util/PkgUtil.java]: Fixed

**Review by zhou9584 — APPROVED** [review]

File: agent/src/main/java/com/microsoft/hydralab/agent/runner/xctest/XctestRunner.java (hunk 4 above)
Comment (lifesaver0129) [reply agent/src/main/java/com/microsoft/hydralab/agent/runner/xctest/XctestRunner.java]: Ok, I will update the doc to declare the dependencies.

**Review by zhou9584 — APPROVED** [review]

**Pull request merged**